dnsclient/dnsclient.go

```diff
@@ -0,0 +1,195 @@
+	if len(rootHash) != sha256.Size {
+		return nil, fmt.Errorf("wrong size root hash %q", matches[3])
+	}
+	signature, err := base64.StdEncoding.DecodeString(matches[4])
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode signature %q", matches[4])
+	}
+
+	sth := ct.SignedTreeHead{
+		TreeSize:  treeSize,
+		Timestamp: timestamp,
+	}
+	copy(sth.SHA256RootHash[:], rootHash)
+
+	var ds ct.DigitallySigned
+	if rest, err := tls.Unmarshal(signature, &ds); err != nil {
+		return nil, fmt.Errorf("failed to parse signature: %v", err)
+	} else if len(rest) > 0 {
+		return nil, fmt.Errorf("trailing data in parse signature: %v", err)
+	}
+	sth.TreeHeadSignature = ds
+	if err := c.verifySTHSignature(sth); err != nil {
+		return nil, fmt.Errorf("signature validation failed: %v", err)
+	}
+	return &sth, nil
+}
+
+// verifySTHSignature checks the signature in sth, returning any error
+// encountered or nil if verification is successful.
+func (c *DNSClient) verifySTHSignature(sth ct.SignedTreeHead) error {
+	if c.Verifier == nil {
+		// Can't verify signatures without a verifier
+		return nil
+	}
+	return c.Verifier.VerifySTHSignature(sth)
+}
+
+// GetSTHConsistency retrieves the consistency proof between two snapshots.
+func (c *DNSClient) GetSTHConsistency(ctx context.Context, first, second uint64) ([][]byte, error) {
+	return c.getProof(ctx, fmt.Sprintf("%d.%d.sth-consistency.%s", first, second, c.base))
+}
+
+// GetProofByHash returns an audit path for the hash of an SCT.
+func (c *DNSClient) GetProofByHash(ctx context.Context, hash []byte, treeSize uint64) (*ct.GetProofByHashResponse, error) {
+	// First get the index of the entry.
+	hash32 := base32.StdEncoding.EncodeToString(hash)
+	// Drop any trailing padding.
+	hash32 = strings.TrimRight(hash32, "=")
+	name := fmt.Sprintf("%s.hash.%s", hash32, c.base)
+	results, err := c.resolve(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("lookup failed for %q: %v", name, err)
+	}
+	result := strings.Join(results, "")
+	leafIndex, err := strconv.ParseUint(result, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse result %q", result)
+	}
+
+	proof, err := c.getProof(ctx, fmt.Sprintf("%d.%d.tree.%s", leafIndex, treeSize, c.base))
+	if err != nil {
+		return nil, err
+	}
+
+	return &ct.GetProofByHashResponse{
+		LeafIndex: int64(leafIndex),
+		AuditPath: proof,
+	}, nil
+}
+
+func (c *DNSClient) getProof(ctx context.Context, base string) ([][]byte, error) {
+	var proof [][]byte
+	for index := 0; index <= 255; {
+		name := fmt.Sprintf("%d.%s", index, base)
+		results, err := c.resolve(ctx, name)
+		if err != nil {
+			if index == 0 {
+				return nil, fmt.Errorf("lookup for %q failed: %v", name, err)
+			}
+			// Assume that a failure to retrieve any more means the proof is complete.
+			break
+		}
+		result := []byte(strings.Join(results, ""))
+		// We expect the result to be a concatenation of hashes, and so should be a multiple of the hash size.
+		if len(result)%sha256.Size != 0 {
```

Comment:
Does it do the right thing if len(result) == 0?

Reply:
Good spot -- the old first==last edge case.  Added a check and a test.